### PR TITLE
AB test: Insert two sets of inline ad slots server-side on liveblogs

### PR DESCRIPTION
**Pull request by domlander**
## What does this change?

- Creates an AB test.
- On liveblog pages, insert two sets of inline ad slots server-side. Fill the relevant slots (mobile or desktop) with ads

## Why?

- We can move from inserting ad slots client side to inserting them server-side. One set of ad slots inserted server-side was not flexible enough to offer a suitable ad frequency on both desktop and mobile. Since viewport height tends to be smaller when the screen width is smaller, number of ads per viewport height is smaller on mobile when inserting ads above the same blocks on all screen sizes.

## Screenshots

Two sets of ad slot containers are created on the page.

| Desktop      | Mobile      |
| ----------- | ---------- |
| The inline ad slots for large screens are filled with ads. The mobile inline ad slots remain unfilled | The ad slots for mobile screens are filled with inline ads. The desktop inline ad slots remain unfilled |
| ![desktop][] | ![mobile][] |

[desktop]: https://github.com/guardian/dotcom-rendering/assets/9574885/1324f279-7780-4b6b-a5fe-425b97deb4bf
[mobile]: https://github.com/guardian/dotcom-rendering/assets/9574885/78353a8a-2533-4be1-b252-0fcf7787a590

### CODE REVIEW

File: dotcom-rendering/src/lib/liveblogAdSlots.ts [hunk 1]
```diff
@@ -68,94 +75,166 @@ type KnownBlockElementType =
...
 const elementHeightDataMap: {
 	[key in KnownBlockElementType]: BlockElementHeightData;
 } = {
 	'model.dotcomrendering.pageElements.BlockquoteBlockElement': {
 		heightExcludingText: 0,
+		heightExcludingTextMobile: 0,
+		margin: 16,
 		textHeight: {
-			lineHeight: 25.5,
+			lineHeight: 23.8,
```
Comment (domlander) [inline dotcom-rendering/src/lib/liveblogAdSlots.ts:88]: These height changes are the result of running a tool in the commercial tools repo that looks to find what inputs can accurately predict the height of an element.

This tool is not yet in the `main` branch of the commercial tools repo, but will be added shortly.

File: dotcom-rendering/src/components/ArticleContainer.tsx [hunk 2]
```diff
@@ -121,19 +120,6 @@ const adStyles = css`
 				z-index: -1;
 			}
 		}
-
```
Comment (domlander) [inline dotcom-rendering/src/components/ArticleContainer.tsx:124]: Moved into `AdSlot.tsx`

File: dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx [hunk 3]
```diff
@@ -59,27 +59,49 @@ export const LiveBlogBlocksAndAdverts = ({
 		);
 	}
 
-	let numPixelsOfContentWithoutAdvert = 0;
-	let numAdvertsInserted = 0;
+	let pxSinceAdMobile = 0;
```
Comment (domlander) [inline dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx:62]: Open to ideas for a cleaner way to do all this

File: dotcom-rendering/src/lib/liveblogAdSlots.ts [hunk 4]
```diff
@@ -20,10 +22,9 @@ const MAX_INLINE_ADS = 8;
 const MIN_SPACE_BETWEEN_ADS = 1_500;
```
Comment (domlander) [inline dotcom-rendering/src/lib/liveblogAdSlots.ts:22]: We had a positive result on desktop with the value in the previous AB test. Ad-ratio up 6%: https://docs.google.com/spreadsheets/d/16DLzkrj4BgyU3iVus2Vd_90z8XGnvsMx1vpljqlZ0H0/edit#gid=745321536

File: dotcom-rendering/src/lib/liveblogAdSlots.ts [hunk 5]
```diff
@@ -170,32 +249,44 @@ export const calculateApproximateElementHeight = (
...
-		const characterCount = heightData.text(element).length;
+	const estimatedHeightWithoutText = heightExcludingText + heightData.margin;
 
-		estimatedHeight += lineHeight * Math.ceil(characterCount / lineLength);
+	if (!heightData.text) {
+		return estimatedHeightWithoutText;
 	}
 
-	return estimatedHeight;
+	// If the element has text that contributes to the height of the element, estimate
```
Comment (emma-imber) [inline dotcom-rendering/src/lib/liveblogAdSlots.ts:262]: All of these comments are really helpful, and make the code much easier to read ✨

File: dotcom-rendering/src/lib/liveblogAdSlots.test.ts [hunk 6]
```diff
@@ -98,6 +173,7 @@ describe('shouldDisplayAd', () => {
 			totalBlocks,
 			numAdsInserted,
 			numPixelsWithoutAdvert,
+			false,
 		);
 
 		expect(result).toBeFalsy();
```
Comment (emma-imber) [inline dotcom-rendering/src/lib/liveblogAdSlots.test.ts:179]: Not a blocker, but is it worth testing this for both mobile and desktop, for consistency with the test for inserting an ad after the first block?

File: dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx [hunk 7]
```diff
@@ -59,27 +59,49 @@ export const LiveBlogBlocksAndAdverts = ({
...
 			{blocks.map((block, i) => {
-				numPixelsOfContentWithoutAdvert +=
-					calculateApproximateBlockHeight(block.elements);
-
-				const willDisplayAdAfterBlock =
+				pxSinceAdMobile += calculateApproximateBlockHeight(
+					block.elements,
+					true,
+				);
+				const willinsertAdMobile =
```
Comment (emma-imber) [inline dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx:75]: Very picky but could we keep consistent camelCase for this var name? `willinsertAdMobile` -> `willInsertAdMobile`

**Review by emma-imber — APPROVED** [review]
Looks good! 🎉 Couple of small things but nothing blocking

File: dotcom-rendering/src/lib/liveblogAdSlots.test.ts (hunk 6 above)
Comment (domlander) [reply dotcom-rendering/src/lib/liveblogAdSlots.test.ts:179]: Yeah good shout, I missed this one. I'll also refactor the file for testing both mobile and desktop as I've remembered a tidier way to do it.

File: dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx (hunk 7 above)
Comment (domlander) [reply dotcom-rendering/src/components/LiveBlogBlocksAndAdverts.tsx:75]: Absolutely! Not picky at all

**Pull request merged**